ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -493,12 +493,40 @@ let vdi_copy_fun __context dbg vdi_map remote is_intra_pool remote_vdis so_far t
       (not is_intra_pool) || (dest_sr_uuid <> vconf.sr)
   in
 
-  let remote_vdi,remote_vdi_reference,newdp,mirror_id =
+
+  (* The following variables (remote_vdi, remote_vdi_reference, newdp, mirror_id) will be
+     gradually filled with actual value along the whole process. By using reference cells,
+     on_failure post processing function can do relevant clean up actions depending on which
+     variables have been initialized which means which steps have happended. *)
+  let remote_vdi = ref "" in
+  let remote_vdi_reference = ref Ref.null in
+  let new_dp = ref None in
+  let mirror_id = ref None in
+
+  let on_failure e =
+    (* Stop mirroring and check to see if it was us that caused the mirror failure *)
+    let mirror_failed =
+      match !mirror_id with
+      | Some mid ->
+         ignore(Storage_access.unregister_mirror mid);
+        (try SMAPI.DATA.MIRROR.stop ~dbg ~id:mid with _ -> ());
+        let m = SMAPI.DATA.MIRROR.stat ~dbg ~id:mid in
+        m.Mirror.failed
+      | None -> false in
+    (* Now that mirroring has finished, clean up any datapath we made *)
+    (match !new_dp with | Some dp -> (try SMAPI.DP.destroy ~dbg ~dp ~allow_leak:false with _ -> info "Failed to cleanup datapath: %s" dp) | None -> ());
+    (* And destroy the new VDI *)
+    if !remote_vdi_reference <> Ref.null then (try XenAPI.VDI.destroy remote.rpc remote.session !remote_vdi_reference with _ -> error "Failed to destroy remote VDI");
+    if mirror_failed then raise (Api_errors.Server_error(Api_errors.mirror_failed,[Ref.string_of vconf.vdi])) else raise e in
+
+  try
     if not mirror then
       let dest_vdi_ref = XenAPI.VDI.get_by_uuid remote.rpc remote.session vdi_uuid in
-      vconf.location,dest_vdi_ref,None,None
+      remote_vdi := vconf.location;
+      remote_vdi_reference := dest_vdi_ref
```

Comment:
This will causes us to try to destroy a remote VDI that wasn't made by us in the exception handler, wont it?